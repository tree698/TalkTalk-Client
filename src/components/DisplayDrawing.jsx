import React from 'react';
import { useNavigate } from 'react-router-dom';
import LoggedInUser from './LoggedInUser';

export default function DisplayDrawing({
  drawing,
  checkboxForDelete,
  toBeDeletedId,
}) {
  const { id, fileName, title, username, brush, description, photo } = drawing;
  const navigate = useNavigate();

  const STYLE_FOR_LOGGEDINUSER =
    'flex items-center justify-center shrink-0 mt-0md:mt-3';
  const STYLE_FOR_AVATAR =
    'w-5 md:w-6 lg:w-8 h-5 md:h-6 lg:h-8 rounded-full mr-0 md:mr-2';
  const STYLE_FOR_USERNAME = 'hidden md:block text-sm';

  return (
    <section className="flex flex-col">
      {checkboxForDelete && (
        <input
          name="delete"
          id="delete"
          onChange={() => toBeDeletedId(id)}
          type="checkbox"
          className="w-3 md:w-4 lg:w-5 h-3 md:h-4 lg:h-5 ml-4 mb-1 border border-lightGray"
        />
      )}
      <li
        onClick={() =>
          navigate('/talk', {
            state: { id, fileName, title, username, brush, description, photo },
          })
        }
        className="list-none rounded-xl shadow-lg cursor-pointer hover:-translate-y-1 hover:brightness-125 transition-all delay-150 duration-300 ease-in-out"
      >
        <img
          src={fileName}
          alt={title}
          className="w-full h-[247px] rounded-t-xl"
        />
        <div className="flex items-center justify-center md:flex-col gap-x-2 md:gap-x-0 p-1 md:p-2 lg:p-3">
          <p className="mb-1 md:mb-2 lg:mb-3 text-sm md:text-lg lg:text-xl font-semibold">
            {title}
          </p>
          <LoggedInUser
            photo={photo}
            username={username}
            styleForAvatar={STYLE_FOR_AVATAR}
            styleForUsername={STYLE_FOR_USERNAME}
            styleForLoggedInUser={STYLE_FOR_LOGGEDINUSER}
          />
        </div>
      </li>
    </section>
  );
}
