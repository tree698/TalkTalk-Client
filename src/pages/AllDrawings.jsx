import React, { useEffect, useState } from 'react';
import { paginationForAllDrawings } from '../config';
import { useApiContext } from '../context/ApiContext';
import DisplayDrawing from '../components/DisplayDrawing';
import ViewMore from '../components/ViewMore';
import toast from 'react-hot-toast';

export default function AllDrawings() {
  const { limit, offset: initialOffset } = paginationForAllDrawings;
  const { workService } = useApiContext();

  const [drawings, setDrawings] = useState([]);
  const [offset, setOffset] = useState(initialOffset);
  const [lengthDrawings, setLengthDrawings] = useState(limit);
  const [error, setError] = useState('');

  useEffect(() => {
    workService
      .getWorks(limit, offset)
      .then((drawings) => setDrawings((prev) => drawings))
      .catch((error) => setError((prev) => error.toString()));
    updateOffset(setOffset, limit, offset);
  }, []);

  useEffect(() => {
    error && toast.error(error);
  }, [error]);

  const handleClick = () => {
    updateOffset(setOffset, limit, offset);
    workService //
      .getWorks(limit, offset)
      .then((drawing) => {
        setLengthDrawings((prev) => drawing.length);
        setDrawings((prev) => [...drawings, ...drawing]);
      })
      .catch((error) => setError((prev) => error.toString()));
  };

  return (
    <section className="flex flex-col items-center flex-1">
      <ul className="max-w-6xl mx-auto grid grid-cols-1 md:grid-cols-3 lg:grid-cols-4 gap-4 p-4">
        {drawings &&
          drawings.map((drawing) => (
            <DisplayDrawing key={drawing.id} drawing={drawing} />
          ))}
      </ul>
      <ViewMore
        lengthDrawings={lengthDrawings}
        limit={limit}
        onButtonClick={handleClick}
      />
    </section>
  );
}

function updateOffset(setOffset, limit, offset) {
  const calculatedOffset = limit + offset;
  return setOffset((prev) => calculatedOffset);
}
