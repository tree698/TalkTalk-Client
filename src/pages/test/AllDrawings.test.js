import renderer from 'react-test-renderer';
import { withApiContext } from '../../test/utils';
import AllDrawings from '../AllDrawings';
import DisplayDrawing from '../../components/DisplayDrawing';
import { fakeDrawings } from '../../test/drawingData';
import { render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

jest.mock('../../components/DisplayDrawing');
jest.mock('../../components/ui/Banner');

describe('AllDrawings', () => {
  const fakeContext = {
    workService: {
      getWorks: jest.fn(),
    },
  };

  afterEach(() => fakeContext.workService.getWorks.mockReset());

  it('renders correctly', () => {
    const component = renderer.create(
      withApiContext(<AllDrawings />, fakeContext)
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('display all drawings', async () => {
    fakeContext.workService.getWorks.mockImplementation(() =>
      Promise.resolve(fakeDrawings)
    );

    DisplayDrawing.mockImplementation((fakeDrawing) => (
      <pre>{JSON.stringify(fakeDrawing)}</pre>
    ));

    await act(async () => {
      render(withApiContext(<AllDrawings />, fakeContext));
    });

    expect(
      screen.getByText(JSON.stringify({ drawing: fakeDrawings[0] }))
    ).toBeInTheDocument();
  });
});
