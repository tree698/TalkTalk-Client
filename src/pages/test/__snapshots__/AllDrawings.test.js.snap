// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllDrawings renders correctly 1`] = `
<section
  className="flex flex-col items-center flex-1"
>
  <ul
    className="w-full grid grid-cols-1 md:grid-cols-3 lg:grid-cols-4 gap-1 md:gap-2 lg:gap-3 px-3 md:px-0"
  />
  <div>
    <button
      className="w-52 md:w-56 lg:w-60 py-1 md:py-2 lg:py-3 my-6 md:my-8 lg:my-10  text-xs md:text-sm lg:text-base font-semibold text-center text-white bg-brand rounded-3xl hover:scale-105 hover:shadow-xl transition-all delay-150 duration-300 ease-in-out"
      onClick={[Function]}
    >
      View More
    </button>
  </div>
</section>
`;
